source/common/upstream/upstream_impl.h

```diff
@@ -850,6 +851,7 @@ class ClusterInfoImpl : public ClusterInfo,
   TransportSocketMatcherPtr socket_matcher_;
   Stats::ScopeSharedPtr stats_scope_;
   mutable ClusterStats stats_;
```

Comment:
so IIRC the game-plan would be to convert this to `Thread::AtomicPtr<ClusterStats> stats_` right? And then `stats()` above would be implemented as `stats_.get()` ?

Reply:
right. more PRs to come. 
The idea is that we only increase when an individual field in the sub-group is accessed (ideally, only when it's inc-ed).

Reply:
I would put the AtomicPtr in this PR, otherwise this PR actually increases memory usage, even in the case where some clusters get no traffic. It will definitely increase memory usage (by a small amount) when all clusters get traffic. But those situations probably don't have a huge  number of clusters.

Reply:
I plan to have another "refactoring" PR before actually introduce the "lazy-init" Struct, to hopefully make the PR easier to read. 
Let me know if you would prefer a bigger PR. :P